source/AI.cpp

```diff
@@ -3313,32 +3313,94 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 		// Board the nearest disabled ship, focusing on hostiles before allies. Holding
 		// `Shift` results in boarding only player-owned escorts in need of assistance.
 		shared_ptr<const Ship> target = ship.GetTargetShip();
-		if(!target || !CanBoard(ship, *target) || (shift && !target->IsYours()))
+		if(!target || (target && activeCommands.Has(Command::WAIT)) || !CanBoard(ship, *target) || (shift && !target->IsYours()))
 		{
 			if(shift)
 				ship.SetTargetShip(shared_ptr<Ship>());
 
-			double closest = numeric_limits<double>::infinity();
 			bool foundEnemy = false;
-			bool foundAnything = false;
-			for(const shared_ptr<Ship> &other : ships)
-				if(CanBoard(ship, *other))
+
+			auto strategy = [&]() -> function<double(const Ship &)> {
+				Point current = ship.Position();
+				// If "board target" is set to true, the default behavior (distance priority) is used.
+				if(!Preferences::Has("Board target"))
 				{
-					if(shift && !other->IsYours())
-						continue;
+					double agility = ship.Acceleration() * ship.TurnRate();
+					return [agility, this, &ship, current](const Ship &other) -> double {
+						double cost = this->Has(ship, other.shared_from_this(), ShipEvent::SCAN_OUTFITS) ?
+							other.Cost() : (other.ChassisCost() * 2.);
+						return -agility * 2. * (cost * cost) / current.DistanceSquared(other.Position());
+					};
+				}
+				// Default to distance-based strategy.
+				return [current](const Ship &other) -> double {
+					return current.DistanceSquared(other.Position());
+				};
+			}();
+
+			auto boardable = vector<pair<const Ship *, double>>{};
+
+			auto fillBoardable = [&ship, &foundEnemy, &boardable, &strategy](const Ship &other)
+			{
+				if(CanBoard(ship, other))
+					boardable.emplace_back(make_pair(&other, strategy(other)));
+			};
 
-					bool isEnemy = other->GetGovernment()->IsEnemy(ship.GetGovernment());
-					double d = other->Position().Distance(ship.Position());
-					if((isEnemy && !foundEnemy) || (d < closest && isEnemy == foundEnemy))
+			if(shift)
+			{
+				for(const Ship *other : governmentRosters[ship.GetGovernment()])
+					if(other->GetSystem() == ship.GetSystem())
+						fillBoardable(*other);
+			}
+			else
+			{
+				// First check if we can board enemy ships, then allies.
+				for(const Ship *other : GetShipsList(ship, true))
+					fillBoardable(*other);
+				// The current target is not considered by GetShipsList.
+				if(target)
+					fillBoardable(*target.get());
+
+				if(boardable.empty())
+					for(const Ship *other : GetShipsList(ship, false))
+						fillBoardable(*other);
+			}
+
+			if(boardable.empty())
+				activeCommands.Clear(Command::BOARD);
+			else
+			{
+				sort(boardable.begin(), boardable.end(),
+					[](
+						pair<const Ship *, double> &lhs,
+						pair<const Ship *, double> &rhs
+					)
 					{
-						closest = d;
-						foundEnemy = isEnemy;
-						foundAnything = true;
-						ship.SetTargetShip(other);
+						return lhs.second > rhs.second;
```

Comment:
Instead of the math do the complexity of the sorting here prioritizing distance over cost within a certain range falling back to cost.

The logic would be the same for both preferences because distance-based is already distance based.

Reply:
I suggest this logic-based approach instead of the math.  https://github.com/Hurleveur/endless-sky/pull/3

Reply:
This can be resolved.